core/src/test/java/org/apache/brooklyn/enricher/stock/reducer/ReducerTest.java

```diff
@@ -0,0 +1,240 @@
+    }
+    
+    @Test
+    public void testReducingBuilderWithJoinerFunctionWithDefaultParameter() {
+        entity.enrichers().add(Enrichers.builder()
+            .reducing(Reducer.class, ImmutableList.<AttributeSensor<?>>of(STR1, STR2))
+            .from(entity)
+            .computing("joiner")
+            .publishing(STR3)
+            .build()
+        );
+        EntityTestUtils.assertAttributeEquals(entity, STR3, null);
+        
+        entity.sensors().set(STR1, "foo");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, null");
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithJoinerFunctionAndUnusedParameter() {
+        
+        entity.enrichers().add(Enrichers.builder()
+            .reducing(Reducer.class, ImmutableList.<AttributeSensor<?>>of(STR1, STR2))
+            .from(entity)
+            .computing("joiner", ImmutableMap.<String, Object>of("non.existent.parameter", "-"))
+            .publishing(STR3)
+            .build()
+        );
+        EntityTestUtils.assertAttributeEquals(entity, STR3, null);
+        
+        entity.sensors().set(STR1, "foo");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, null");
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithFormatStringFunction() {
+        entity.enrichers().add(Enrichers.builder()
+            .reducing(Reducer.class, ImmutableList.<AttributeSensor<?>>of(STR1, STR2))
+            .from(entity)
+            .computing("formatString", ImmutableMap.<String, Object>of("format", "hello, %s and %s"))
+            .publishing(STR3)
+            .build()
+        );
+        
+        EntityTestUtils.assertAttributeEquals(entity, STR3, null);
+        
+        entity.sensors().set(STR1, "foo");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "hello, foo and null");
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "hello, foo and bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithNamedNonExistentFunction() {
+        try { 
+            entity.enrichers().add(Enrichers.builder()
+                .reducing(Reducer.class, ImmutableList.<AttributeSensor<?>>of(STR1, STR2))
+                .from(entity)
+                .computing("unknown function name", ImmutableMap.<String, Object>of("separator", "-"))
+                .publishing(STR3)
+                .build()
+            );
+            Asserts.fail("Expected exception when adding reducing enricher with unknown named function");
+        } catch (Exception e) {
+            Throwable t = Exceptions.getFirstThrowableOfType(e, IllegalStateException.class);
```

Comment:
I much prefer:

    if (Exceptions.getFirstThrowableOfType(e, IllegalStateException.class) == null) {
        throw e;
    }

Otherwise we lose the actual exception. This is normally ok, but it's a real pain if a test fails non-deterministically in the asfbot with no information about what went wrong.